Apromore-Extras/APMLogModule/src/main/java/org/apromore/apmlog/logobjects/AbstractLogImpl.java

```diff
@@ -101,4 +104,8 @@ protected void setActivityInstances(List<ActivityInstance> activityInstances) {
     public void setTimeZone(String timeZone) {
         this.timeZone = timeZone;
     }
+
+    public void setCalendarModel(CalendarModel calendarModel) {
+        this.calendarModel = calendarModel;
+    }
```

Comment:
It's better to add a @NonNull constraint, e.g. setCalendarModel(@NonNull CalendarModel calendarModel). It will help the debugging of NullPointerException because it will be thrown right in this set method if it happens rather than waiting until some methods encounter null and throw it.

Reply:
Another comment is it's safer to set CalendarModel in the constructor rather than via setCalendarModel method whiich is separated from log constructor and can be forgotten to call. Without a calendar model, all duration computation for the log won't work, that's the reason why it's safer to set calendar model in APMLog constructor  and check for non-null. With it initialized in the constructor, Once APMLog has been created, it is always safe to use its calendar model, no need to check for calendarModel != null condition in many places in the code.

Reply:
If possible, marking `calendarModel` as `final` would allow the compiler to confirm that it's not changed anywhere other than in the constructor.

Reply:
both comments are not applicable.

Reply:
making calendarModel final is not applicable

Reply:
That's ok. I remembered now APMLog is cached when it's not associated with any calendar.